src/v/wasm/wasmtime.cc

```diff
@@ -0,0 +1,759 @@
+                sr_module.get(), linker.get(), alien, shard_id));
+
+              std::vector<ss::sstring> args{_meta.name()};
+              absl::flat_hash_map<ss::sstring, ss::sstring> env
+                = _meta.environment;
+              env.emplace("REDPANDA_INPUT_TOPIC", _meta.input_topic.tp());
+              env.emplace(
+                "REDPANDA_OUTPUT_TOPIC", _meta.output_topics.begin()->tp());
+              auto wasi_module = std::make_unique<wasi::preview1_module>(
+                std::move(args), std::move(env), _logger);
+              host_function_envs.push_back(register_wasi_module(
+                wasi_module.get(), linker.get(), alien, shard_id));
+
+              wasmtime_instance_t instance;
+              wasm_trap_t* trap_ptr = nullptr;
+              error.reset(wasmtime_linker_instantiate(
+                linker.get(), context, _module.get(), &instance, &trap_ptr));
+              handle<wasm_trap_t, wasm_trap_delete> trap{trap_ptr};
+              check_error(error.get());
+              check_trap(trap.get());
+
+              return std::make_unique<wasmtime_engine>(
+                _meta.name(),
+                std::move(linker),
+                std::move(store),
+                _module,
+                std::move(xform_module),
+                std::move(sr_module),
+                std::move(wasi_module),
+                std::move(host_function_envs),
+                instance,
+                _workers);
+          });
+    }
+
+private:
+    wasm_engine_t* _engine;
+    std::shared_ptr<wasmtime_module_t> _module;
+    model::transform_metadata _meta;
+    schema_registry* _sr;
+    ss::logger* _logger;
+    ssx::sharded_thread_worker* _workers;
+};
+
+class wasmtime_runtime : public runtime {
+public:
+    wasmtime_runtime(
+      handle<wasm_engine_t, &wasm_engine_delete> h,
+      std::unique_ptr<schema_registry> sr)
+      : _engine(std::move(h))
+      , _sr(std::move(sr)) {}
+
+    ss::future<> start() override {
+        co_await _workers.start({.name = "wasm"});
+        co_await ss::smp::invoke_on_all([this] {
+            return _workers.submit([] {
+                // wasmtime needs some signals for it's handling, make sure we
+                // unblock them
+                auto mask = ss::make_empty_sigset_mask();
+                sigaddset(&mask, SIGSEGV);
+                sigaddset(&mask, SIGILL);
+                sigaddset(&mask, SIGFPE);
+                ss::throw_pthread_error(
+                  ::pthread_sigmask(SIG_UNBLOCK, &mask, nullptr));
+            });
+        });
+    }
+
+    ss::future<> stop() override { return _workers.stop(); }
+
+    ss::future<std::unique_ptr<factory>> make_factory(
+      model::transform_metadata meta, iobuf buf, ss::logger* logger) override {
+        auto user_module = co_await _workers.submit([this, &meta, &buf] {
+            vlog(wasm_log.debug, "compiling wasm module {}", meta.name);
+            bytes b = iobuf_to_bytes(buf);
```

Comment:
this serializes the buffer. will it be big and risk OOM when memory is fragmented?

Reply:
This happens in an alien thread, which uses the system allocator, not the seastar one. I added a comment to that effect.

Reference: https://github.com/redpanda-data/seastar/blob/ee8c8da17ec6108dcee090a33c11717a6d42896e/src/core/memory.cc#L1616

Reply:
>This happens in an alien thread, which uses the system allocator, not the seastar one. I added a comment to that effect.

sgtm. i wish there was some way to mark clearly code on alien threads with a convention. well, we don't have that many cases so it's not a huge deal either way.

Reply:
I can rename the varible to `alien_thread_pool` so it makes this a bit more obvious (along with the comment).

Reply:
Done.